### PR TITLE
Numeric input interaction view validation

### PR DESCRIPTION
**Pull request by nikitamarysolomanpvt**
## Explanation
Added validation for number input with error messages given below.
    INVALID_CHARS = "Please only use numerical digits, minus(-) or dot (.)"
    INVALID_FORMAT = "Please enter a valid number"
    STARTING_WITH_FLOATING_POINT= "Please begin your answer with a number (e.g.,”0” in 0.5)"

For testing purpose make following change in code:

Launcher activity should be InputInteractionViewTestActivity

## Design Doc

https://docs.google.com/document/d/1QigNdZeMBy-JjHW26QH4qT0UtCyzfvW0xEaStSdY9gM/edit#heading=h.2i0txgpugkgf

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/res/layout/activity_input_interaction_view_test.xml
Comment (nikitamarysolomanpvt) [inline app/src/main/res/layout/activity_input_interaction_view_test.xml:38]: Removed for testing purpose.

File: app/src/main/res/layout/activity_input_interaction_view_test.xml
Comment (rt4914) [inline app/src/main/res/layout/activity_input_interaction_view_test.xml:80]: `android:inputType="numberSigned|numberDecimal"` 

This attribute helps to make sure that the number keyboard is shown in which you can use following characters, `-`, `.` and `0123456789` Also, it handles a lot of things on it own. Try using this, it might help more.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Remove unused imports

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Update this to more kotlin style:
```
      pendingAnswerError = when (category) {
        AnswerErrorCategory.REAL_TIME -> getRealTimeAnswerError(answerText.toString()).getErrorMessageFromStringRes(
          context
        )
        AnswerErrorCategory.SUBMIT_TIME -> getSubmitTimeError(answerText.toString()).getErrorMessageFromStringRes(
          context
        )
      }
```

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Typo here in at other places to `NumericInpurParsingError`, should be replaced with `NumericInputParsingError`

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Remove unused import.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: ditto.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Why is there `--` at the end of this sentence?

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: This should be private or is it getting used somewhere which is not getting detected in my android studio.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: This should be private or is it getting used somewhere which is not getting detected in my android studio.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Make this function private.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Ditto.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Typo error make it NumericInputParsingError().

**Review by rt4914 — CHANGES_REQUESTED** [review]
@nikitamarysolomanpvt The code looks good. I have suggested one change to `NumericInputInteractionView` in xml file. Try that out. 
Also, currently, `5.` is a valid number, do we intend to keep it that way?

Also, 3 test cases are failing in my laptop but the error it is giving me is that the string-resource not found but actually those resources are available. So I will need to defer to @veena14cs regarding the test cases.

Screenshot 2020-02-06 at 11 16 12 AM

**Review by veena14cs — APPROVED** [review]
LGTM besides some nit change and add full stop to all the error strings in strings.xml.

**Comment by veena14cs** [comment]
> @nikitamarysolomanpvt The code looks good. I have suggested one change to `NumericInputInteractionView` in xml file. Try that out.
> Also, currently, `5.` is a valid number, do we intend to keep it that way?
> 
> Also, 3 test cases are failing in my laptop but the error it is giving me is that the string-resource not found but actually those resources are available. So I will need to defer to @veena14cs regarding the test cases.
> 
> Screenshot 2020-02-06 at 11 16 12 AM

I am able to run all the testcases successfully.

**Comment by rt4914** [comment]
> > @nikitamarysolomanpvt The code looks good. I have suggested one change to `NumericInputInteractionView` in xml file. Try that out.
> > Also, currently, `5.` is a valid number, do we intend to keep it that way?
> > Also, 3 test cases are failing in my laptop but the error it is giving me is that the string-resource not found but actually those resources are available. So I will need to defer to @veena14cs regarding the test cases.
> > Screenshot 2020-02-06 at 11 16 12 AM
> 
> I am able to run all the testcases successfully.

Yeah, that is what I was also expecting. So its an android studio issue.

**Comment by veena14cs** [comment]
> > > @nikitamarysolomanpvt The code looks good. I have suggested one change to `NumericInputInteractionView` in xml file. Try that out.
> > > Also, currently, `5.` is a valid number, do we intend to keep it that way?
> > > Also, 3 test cases are failing in my laptop but the error it is giving me is that the string-resource not found but actually those resources are available. So I will need to defer to @veena14cs regarding the test cases.
> > > Screenshot 2020-02-06 at 11 16 12 AM
> > 
> > 
> > I am able to run all the testcases successfully.
> 
> Yeah, that is what I was also expecting. So its an android studio issue.

May be yes.

File: app/src/main/res/layout/activity_input_interaction_view_test.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/layout/activity_input_interaction_view_test.xml]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done

File: app/src/main/res/layout/activity_input_interaction_view_test.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/layout/activity_input_interaction_view_test.xml:80]: This was suggested by ben in fraction input to keep it open for testing purpose.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Removed

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done

File: app/src/main/res/values/strings.xml
Comment (rt4914) [inline app/src/main/res/values/strings.xml]: there is an extra space after `...number`. Remove this.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

File: app/src/main/res/values/strings.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Please make this a ``val`` since it doesn't actually change, just the underlying observable value does.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt:17]: Can you move the error checking logic into its own class like it is for fractions? I think that will make the view model a bit easier to read & understand since it keeps it more focused (e.g. the view model should only care about binding data & arranging stuff, whereas the error checker specifically cares about computing an error rather than having one class that does both).

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: This & the next name are misleading. Isn't this pattern for checking valid characters? E.g. if there are any characters other than these, the string contains an invalid character? Ditto for the length pattern below.

Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Also to confirm: we do not want to allow scientific notation, correct?

File: app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt:57]: Nit for consistency on order & elsewhere in the PR for interactions: please consistently make fractions before or after numeric input rather than switching them.

File: app/src/main/res/layout/numeric_input_interaction_item.xml
Comment (BenHenning) [inline app/src/main/res/layout/numeric_input_interaction_item.xml]: Nit: ``viewModel.errorMessage.get().length() > 0 ? View.VISIBLE : View.INVISIBLE`` (spacing between the condition operators). Ditto also above in the test version of this layout.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: Space after ``number``.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: Is this the final wording @mschanteltc suggested? It feels a bit context-less. Should we say something like: "The answer can contain at most 15 numbers." or something similar? Also, do users know what 'character' means in this context?

File: app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt]: Maybe also test the case where the number is too long without a decimal.

File: app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt:277]: Did you test the case where the number is successfully parsed? If not, please add. Also, it might be nice to ones for decimals & negatives, as well (e.g. ~3 success test cases).

**Review by BenHenning — APPROVED** [review]
Thanks @nikitamarysolomanpvt! Generally LGTM, though please make sure the open comments are fully resolved. I'll approve, but consider sending back to me if you'd like me to do one more pass before submission.

File: app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt]: @BenHenning We are not switching them we are calling the checkPendingAnswerError for the respective inputs from their viewModels... Could you please elaborate if you are not speak about the same.

File: app/src/main/res/values/strings.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: @BenHenning In my opinion digit would be a familiar term to kids i.e. "The answer can contain at most 15 digits."

Comment (mschanteltc) [reply app/src/main/res/values/strings.xml]: If we allow at most 15 "numbers," would this mean the decimal point (and possibly forward slashes, negative signs, etc.) would be excluded from the count? I agree "characters" can sound a bit out of context. "Characters" was used to describe all text input.

Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: Is this ok? "The answer can contain at most 15 characters(i.e.,[0-9] or "."or "-")"

Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: > If we allow at most 15 "numbers," would this mean the decimal point (and possibly forward slashes, negative signs, etc.) would be excluded from the count? I agree "characters" can sound a bit out of context. "Characters" was used to describe all text input.

@mschanteltc 
only "."and "-" are allowed, no other characters are allowed in number input.

File: app/src/main/res/values/strings.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: Done.

File: app/src/main/res/layout/numeric_input_interaction_item.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/layout/numeric_input_interaction_item.xml]: Done

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Yes, we do not want to allow scientific notation.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt:17]: Separated to StringToNumberParser file

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: updated to " validCharsRegex"

File: app/src/main/res/values/strings.xml
Comment (mschanteltc) [reply app/src/main/res/values/strings.xml]: @seanlip Will other answer inputs (such as words and equations) follow this 15 character max as well? If so, I think we should stick with the word "character" out of consistency.

File: app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt
Comment (nikitamarysolomanpvt) [reply app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt]: Ok Done

File: app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt
Comment (nikitamarysolomanpvt) [reply app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt:277]: It was already there. added new test for negative number.

File: app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt:57]: Done. Changed to Numeric first then Fraction Interaction.

Comment (rt4914) [reply app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt:57]: @nikitamarysolomanpvt Suggest keep it other way around. fraction first and then numeric considering alphabetical order.

**Review by rt4914 — APPROVED** [review]
LGTM again. Thanks

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Is this also true on Oppia web?

Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: I was not able to  find any parser for scientific notation in numeric input of oppia web code base.@Sean could you please let me know if iam wrong about the same or if we need to allow scientific notation in android?.

Comment (seanlip) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Please use @seanlip, not @sean. The latter does not reach me.

We do not have a scientific notation parser in the web codebase and we do not plan to support it.

File: app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/testing/InputInteractionViewTestActivity.kt:57]: Done.

File: app/src/main/res/values/strings.xml
Comment (nikitamarysolomanpvt) [reply app/src/main/res/values/strings.xml]: @seanlip PTAL on this and suggest the error message for number too long, currently we are showing "The answer can contain at most 15 characters (i.e.,[0-9] or "."or "-")."

**Review by rt4914 — COMMENTED** [review]
I have checked the alphabetical order of interactions- fraction, numeric, text. It looks correct to me.

File: app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt]: Okay thank you.

**Comment by nikitamarysolomanpvt** [comment]
We are merging this to unblock number with units. 
TODO
Numeric input -"number too long" validation error message change #652

File: app/src/main/res/values/strings.xml
Comment (seanlip) [reply app/src/main/res/values/strings.xml]: @mschanteltc Sorry, missed your question. Other answer inputs will not have this maximum limit.

Perhaps: "The answer can contain at most 15 digits (0-9) or symbols (. or -)."

**Pull request merged**
